Test failures on s390x

I tried building and running the tests on Fedora/s390x and got a lot of test failures.  From the test output, a lot of the bytes appear reversed, maybe this is an issue of big vs. little endianess?
Full build info is in koji:
https://koji.fedoraproject.org/koji/taskinfo?taskID=26880670
https://kojipkgs.fedoraproject.org//work/tasks/670/26880670/build.log

```
--- FAIL: TestMarshalAttributes (0.00s)
    --- FAIL: TestMarshalAttributes/one_attribute,_no_data (0.00s)
    	attribute_test.go:263: unexpected bytes:
    		- want: [0x04 0x00 0x01 0x00]
    		-  got: [0x00 0x04 0x00 0x01]
    --- FAIL: TestMarshalAttributes/one_attribute,_no_data,_length_calculated (0.00s)
    	attribute_test.go:263: unexpected bytes:
    		- want: [0x04 0x00 0x01 0x00]
    		-  got: [0x00 0x04 0x00 0x01]
    --- FAIL: TestMarshalAttributes/one_attribute,_padded (0.00s)
    	attribute_test.go:263: unexpected bytes:
    		- want: [0x05 0x00 0x01 0x00 0xff 0x00 0x00 0x00]
    		-  got: [0x00 0x05 0x00 0x01 0xff 0x00 0x00 0x00]
    --- FAIL: TestMarshalAttributes/one_attribute,_padded,_length_calculated (0.00s)
    	attribute_test.go:263: unexpected bytes:
    		- want: [0x05 0x00 0x01 0x00 0xff 0x00 0x00 0x00]
    		-  got: [0x00 0x05 0x00 0x01 0xff 0x00 0x00 0x00]
    --- FAIL: TestMarshalAttributes/one_attribute,_aligned (0.00s)
    	attribute_test.go:263: unexpected bytes:
    		- want: [0x08 0x00 0x02 0x00 0xaa 0xbb 0xcc 0xdd]
    		-  got: [0x00 0x08 0x00 0x02 0xaa 0xbb 0xcc 0xdd]
    --- FAIL: TestMarshalAttributes/one_attribute,_aligned,_length_calculated (0.00s)
    	attribute_test.go:263: unexpected bytes:
    		- want: [0x08 0x00 0x02 0x00 0xaa 0xbb 0xcc 0xdd]
    		-  got: [0x00 0x08 0x00 0x02 0xaa 0xbb 0xcc 0xdd]
    --- FAIL: TestMarshalAttributes/multiple_attributes (0.00s)
    	attribute_test.go:263: unexpected bytes:
    		- want: [0x05 0x00 0x01 0x00 0xff 0x00 0x00 0x00 0x08 0x00 0x02 0x00 0xaa 0xbb 0xcc 0xdd 0x04 0x00 0x03 0x00 0x10 0x00 0x04 0x00 0x11 0x11 0x11 0x11 0x22 0x22 0x22 0x22 0x33 0x33 0x33 0x33]
    		-  got: [0x00 0x05 0x00 0x01 0xff 0x00 0x00 0x00 0x00 0x08 0x00 0x02 0xaa 0xbb 0xcc 0xdd 0x00 0x04 0x00 0x03 0x00 0x10 0x00 0x04 0x11 0x11 0x11 0x11 0x22 0x22 0x22 0x22 0x33 0x33 0x33 0x33]
    --- FAIL: TestMarshalAttributes/multiple_attributes,_length_calculated (0.00s)
    	attribute_test.go:263: unexpected bytes:
    		- want: [0x05 0x00 0x01 0x00 0xff 0x00 0x00 0x00 0x08 0x00 0x02 0x00 0xaa 0xbb 0xcc 0xdd 0x04 0x00 0x03 0x00 0x10 0x00 0x04 0x00 0x11 0x11 0x11 0x11 0x22 0x22 0x22 0x22 0x33 0x33 0x33 0x33]
    		-  got: [0x00 0x05 0x00 0x01 0xff 0x00 0x00 0x00 0x00 0x08 0x00 0x02 0xaa 0xbb 0xcc 0xdd 0x00 0x04 0x00 0x03 0x00 0x10 0x00 0x04 0x11 0x11 0x11 0x11 0x22 0x22 0x22 0x22 0x33 0x33 0x33 0x33]
    --- FAIL: TestMarshalAttributes/nested_bit,_type_1,_length_0 (0.00s)
    	attribute_test.go:263: unexpected bytes:
    		- want: [0x04 0x00 0x01 0x80]
    		-  got: [0x00 0x04 0x80 0x01]
    --- FAIL: TestMarshalAttributes/endianness_bit,_type_1,_length_0 (0.00s)
    	attribute_test.go:263: unexpected bytes:
    		- want: [0x04 0x00 0x01 0x40]
    		-  got: [0x00 0x04 0x40 0x01]
    --- FAIL: TestMarshalAttributes/max_type_space,_length_0 (0.00s)
    	attribute_test.go:263: unexpected bytes:
    		- want: [0x04 0x00 0xff 0x3f]
    		-  got: [0x00 0x04 0x3f 0xff]
--- FAIL: TestUnmarshalAttributes (0.00s)
    --- FAIL: TestUnmarshalAttributes/one_attribute,_no_data (0.00s)
    	attribute_test.go:469: unexpected error:
    		- want: <nil>
    		-  got: invalid attribute; length too short or too large
    --- FAIL: TestUnmarshalAttributes/one_attribute,_padded (0.00s)
    	attribute_test.go:469: unexpected error:
    		- want: <nil>
    		-  got: invalid attribute; length too short or too large
    --- FAIL: TestUnmarshalAttributes/one_attribute,_aligned (0.00s)
    	attribute_test.go:469: unexpected error:
    		- want: <nil>
    		-  got: invalid attribute; length too short or too large
    --- FAIL: TestUnmarshalAttributes/multiple_attributes (0.00s)
    	attribute_test.go:469: unexpected error:
    		- want: <nil>
    		-  got: invalid attribute; length too short or too large
    --- FAIL: TestUnmarshalAttributes/nested_and_endianness_bits (0.00s)
    	attribute_test.go:469: unexpected error:
    		- want: invalid attribute; type cannot have both nested and net byte order flags
    		-  got: invalid attribute; length too short or too large
    --- FAIL: TestUnmarshalAttributes/nested_bit,_type_1,_length_0 (0.00s)
    	attribute_test.go:469: unexpected error:
    		- want: <nil>
    		-  got: invalid attribute; length too short or too large
    --- FAIL: TestUnmarshalAttributes/endianness_bit,_type_1,_length_0 (0.00s)
    	attribute_test.go:469: unexpected error:
    		- want: <nil>
    		-  got: invalid attribute; length too short or too large
    --- FAIL: TestUnmarshalAttributes/max_type_space,_length_0 (0.00s)
    	attribute_test.go:469: unexpected error:
    		- want: <nil>
    		-  got: invalid attribute; length too short or too large
--- FAIL: TestMessageMarshal (0.00s)
    --- FAIL: TestMessageMarshal/OK_no_data (0.00s)
    	message_test.go:228: unexpected Message bytes:
    		- want: [0x10 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00]
    		-  got: [0x00 0x00 0x00 0x10 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00]
    --- FAIL: TestMessageMarshal/OK_unaligned_data (0.00s)
    	message_test.go:228: unexpected Message bytes:
    		- want: [0x14 0x00 0x00 0x00 0x00 0x00 0x01 0x00 0x01 0x00 0x00 0x00 0x0a 0x00 0x00 0x00 0x61 0x62 0x63 0x00]
    		-  got: [0x00 0x00 0x00 0x14 0x00 0x00 0x00 0x01 0x00 0x00 0x00 0x01 0x00 0x00 0x00 0x0a 0x61 0x62 0x63 0x00]
    --- FAIL: TestMessageMarshal/OK_aligned_data (0.00s)
    	message_test.go:228: unexpected Message bytes:
    		- want: [0x14 0x00 0x00 0x00 0x02 0x00 0x00 0x00 0x02 0x00 0x00 0x00 0x14 0x00 0x00 0x00 0x61 0x62 0x63 0x64]
    		-  got: [0x00 0x00 0x00 0x14 0x00 0x02 0x00 0x00 0x00 0x00 0x00 0x02 0x00 0x00 0x00 0x14 0x61 0x62 0x63 0x64]
--- FAIL: TestMessageUnmarshal (0.00s)
    --- FAIL: TestMessageUnmarshal/OK_no_data (0.00s)
    	message_test.go:309: unexpected error:
    		- want: <nil>
    		-  got: not enough data to create a netlink message
    --- FAIL: TestMessageUnmarshal/OK_data (0.00s)
    	message_test.go:309: unexpected error:
    		- want: <nil>
    		-  got: not enough data to create a netlink message
--- FAIL: TestConnReceiveErrorLinux (0.00s)
    --- FAIL: TestConnReceiveErrorLinux/ENOENT (0.00s)
    	conn_linux_error_test.go:87: unexpected error:
    		- want: no such file or directory
    		-  got: errno 16777217
    --- FAIL: TestConnReceiveErrorLinux/multipart_done_with_error_attached (0.00s)
    	conn_linux_error_test.go:87: unexpected error:
    		- want: interrupted system call
    		-  got: errno 50331649
```

--- a/attribute_test.go
+++ b/attribute_test.go
@@ -7,6 +7,8 @@ import (
 )
 
 func TestMarshalAttributes(t *testing.T) {
+	skipBigEndian(t)
+
 	tests := []struct {
 		name  string
 		attrs []Attribute
@@ -268,6 +270,8 @@ func TestMarshalAttributes(t *testing.T) {
 }
 
 func TestUnmarshalAttributes(t *testing.T) {
+	skipBigEndian(t)
+
 	tests := []struct {
 		name  string
 		b     []byte

--- a/conn_linux_error_test.go
+++ b/conn_linux_error_test.go
@@ -3,14 +3,18 @@
 package netlink_test
 
 import (
+	"encoding/binary"
 	"testing"
 
 	"github.com/mdlayher/netlink"
+	"github.com/mdlayher/netlink/nlenc"
 	"github.com/mdlayher/netlink/nltest"
 	"golang.org/x/sys/unix"
 )
 
 func TestConnReceiveErrorLinux(t *testing.T) {
+	skipBigEndian(t)
+
 	// Note: using *Conn instead of Linux-only *conn, to test
 	// error handling logic in *Conn.Receive
 
@@ -88,5 +92,11 @@ func TestConnReceiveErrorLinux(t *testing.T) {
 					want, got)
 			}
 		})
+	}
+}
+
+func skipBigEndian(t *testing.T) {
+	if nlenc.NativeEndian() == binary.BigEndian {
+		t.Skip("skipping test on big-endian system")
 	}
 }

--- a/message_test.go
+++ b/message_test.go
@@ -2,11 +2,12 @@ package netlink
 
 import (
 	"bytes"
+	"encoding/binary"
 	"reflect"
 	"testing"
-)
 
-// TODO(mdlayher): tests all assume little endian host machine
+	"github.com/mdlayher/netlink/nlenc"
+)
 
 func TestHeaderFlagsString(t *testing.T) {
 	tests := []struct {
@@ -127,6 +128,8 @@ func TestHeaderTypeString(t *testing.T) {
 }
 
 func TestMessageMarshal(t *testing.T) {
+	skipBigEndian(t)
+
 	tests := []struct {
 		name string
 		m    Message
@@ -232,6 +235,8 @@ func TestMessageMarshal(t *testing.T) {
 }
 
 func TestMessageUnmarshal(t *testing.T) {
+	skipBigEndian(t)
+
 	tests := []struct {
 		name string
 		b    []byte
@@ -449,5 +454,11 @@ func TestValidate(t *testing.T) {
 					want, got)
 			}
 		})
+	}
+}
+
+func skipBigEndian(t *testing.T) {
+	if nlenc.NativeEndian() == binary.BigEndian {
+		t.Skip("skipping test on big-endian system")
 	}
 }

--- a/nlenc/endian.go
+++ b/nlenc/endian.go
@@ -1,0 +1,16 @@
+package nlenc
+
+import (
+	"encoding/binary"
+)
+
+// NativeEndian returns the native byte order of this system.
+func NativeEndian() binary.ByteOrder {
+	// Determine endianness by storing a uint16 in a byte slice.
+	b := Uint16Bytes(1)
+	if b[0] == 1 {
+		return binary.LittleEndian
+	}
+
+	return binary.BigEndian
+}

